File: Day_006/majority-element.py
class Solution:
    def findMajority(self, arr):
        l = len(arr)
        e1 ,e2 ,c1 ,c2 = -1 ,-1 ,0 ,0
        
        for num in arr :
            if num == e1:
                c1 += 1
            
            elif num == e2:
                c2 += 1
                
            elif c1 == 0:
                e1 = num
                c1 += 1
                
            elif c2 == 0:
                e2 = num
                c2 += 1
                
            else :
                c1 -= 1;
                c2 -= 1;
                
        res = []
        ct1 ,ct2 = 0 ,0
        
        for n in arr :
            if n == e1:
                ct1 += 1
            
            if n == e2:
                ct2 += 1
                
        if (ct1 > l/3):
            res.append(e1)
            
        if (ct2 > l/3):
            res.append(e2)
            
        res.sort()
        return res

File: Day_006/majority-element.java
import java.util.*;

class Solution {
    public ArrayList<Integer> findMajority(int[] arr) {
        int l = arr.length;
        int e1 = -1, e2 = -1, c1 = 0, c2 = 0;
        
        for (int num : arr) {
            if (num == e1){
                c1 += 1;
            }
            else if (num == e2) {
                c2 += 1;
            }
            else if (c1 == 0) {
                e1 = num;
                c1 += 1;
            }
            else if (c2 == 0) {
                e2 = num;
                c2 += 1;
            }
            else {
                c1--;
                c2--;
            }
        }
        
        ArrayList<Integer> res = new ArrayList<>();
        int ct1 = 0, ct2 = 0;
        
        for (int num : arr) {
            if (e1 == num) {
                ct1++;
            }
            else if (e2 == num) {
                ct2++;
            }
        }
        
        if (ct1 > l/3){
            res.add(e1);
        }
        
        if (ct2 > l/3 && e1 != e2) {
            res.add(e2);
        }
        
        Collections.sort(res);
        return (res);
    }
}